--- ./bl2/bl2_image_load_v2.c
+++ ./bl2/bl2_image_load_v2.c
@@ -74,9 +74,21 @@
 			err = load_auth_image(bl2_node_info->image_id,
 				bl2_node_info->image_info);
 			if (err != 0) {
-				ERROR("BL2: Failed to load image id %u (%i)\n",
+				ERROR("BL2: Failed to load image id %u (%i). Retrying (1)\n",
 				      bl2_node_info->image_id, err);
-				plat_error_handler(err);
+	    		err = load_auth_image(bl2_node_info->image_id,
+    				bl2_node_info->image_info);
+    			if (err != 0) {
+    				ERROR("BL2: Failed to load image id %u (%i). Retrying (2)\n",
+    				      bl2_node_info->image_id, err);
+			        err = load_auth_image(bl2_node_info->image_id,
+        				bl2_node_info->image_info);
+		        	if (err != 0) {
+		        		ERROR("BL2: Failed to load image id %u (%i)\n",
+		        		      bl2_node_info->image_id, err);
+		        		plat_error_handler(err);
+        			}
+    			}
 			}
 		} else {
 			INFO("BL2: Skip loading image id %u\n", bl2_node_info->image_id);
